Framework/API/src/ExperimentInfo.cpp

```diff
@@ -204,6 +213,12 @@ void ExperimentInfo::setInstrument(const Instrument_const_sptr &instr) {
     m_parmap = boost::make_shared<ParameterMap>();
   }
   m_detectorInfo = makeDetectorInfo(*sptr_instrument, *instr);
+  // Detector IDs that were previously dropped because they were not part of the
+  // instrument may now suddenly be valid, so we have to reinitialize the
+  // detector grouping. Also the index corresponding to specific IDs may have
+  // changed.
+  std::fill(m_spectrumDefinitionNeedsUpdate.begin(),
```

Comment:
I think your storage type for the `m_spectrumDefinitionNeedsUpdate` is good. However i think it might make things slightly more readable of updates to that vector were made via some better named methods. So in this case something like `invalidateAllSpectrumDefinitions()`.